--- a/ChangeLog
+++ b/ChangeLog
@@ -122,6 +122,14 @@ Mon May 21 16:29:47 2012  Akinori MUSHA
 	* ext/syslog/syslog.c (mSyslog_inspect): Make sure self is a
 	  module before calling rb_class2name().
 
+Sat May 19 14:42:20 2012  Tanaka Akira  <akr@fsij.org>
+
+	* lib/drb/ssl.rb: generate 1024 bits RSA key instead of 512 bits.
+	  OpenSSL 1.0.1 rejects 512 bits RSA key for TLS1.2 with SHA512.
+	  http://rt.openssl.org/Ticket/Display.html?id=2769&user=guest&pass=guest
+	  reported by Bohuslav Kabrda.
+	  [ruby-core:43844] [ruby-trunk - Bug #6221]
+
 Fri May 11 14:09:48 2012  Nobuyoshi Nakada  <nobu@ruby-lang.org>
 
 	* ext/bigdecimal/bigdecimal.c (PUSH): to prevent VALUE from GC,
--- a/lib/drb/ssl.rb
+++ b/lib/drb/ssl.rb
@@ -54,7 +54,7 @@ module DRb
           return
         end
 
-	rsa = OpenSSL::PKey::RSA.new(512){|p, n|
+	rsa = OpenSSL::PKey::RSA.new(1024){|p, n|
 	  next unless self[:verbose]
 	  case p
 	  when 0; $stderr.putc "."  # BN_generate_prime
