diff --git a/version.h b/version.h
index 6b6d57e..38a9d7b 100644
--- a/version.h
+++ b/version.h
@@ -1,8 +1,8 @@
 #define RUBY_VERSION "1.8.7"
-#define RUBY_RELEASE_DATE "2013-06-27"
+#define RUBY_RELEASE_DATE "2015-04-14"
 #define RUBY_VERSION_CODE 187
-#define RUBY_RELEASE_CODE 20130627
-#define RUBY_PATCHLEVEL 374
+#define RUBY_RELEASE_CODE 20150414
+#define RUBY_PATCHLEVEL 375
 
 #define RUBY_VERSION_MAJOR 1
 #define RUBY_VERSION_MINOR 8
